// igniteui-angular/esm2015/lib/grids/filtering/excel-style/excel-style-date-expression.component.js
import { Component, ViewChild, Input } from '@angular/core';
import { IgxExcelStyleDefaultExpressionComponent } from './excel-style-default-expression.component';
import { IgxDatePickerComponent } from '../../../date-picker/date-picker.component';
/**
 * @hidden
 */
import * as ɵngcc0 from '@angular/core';
import * as ɵngcc1 from '../../../select/select.component';
import * as ɵngcc2 from '../../../directives/prefix/prefix.directive';
import * as ɵngcc3 from '@angular/common';
import * as ɵngcc4 from '../../../date-picker/date-picker.component';
import * as ɵngcc5 from '@angular/forms';
import * as ɵngcc6 from '../../../date-picker/date-picker.directives';
import * as ɵngcc7 from '../../../directives/toggle/toggle.directive';
import * as ɵngcc8 from '../../../icon/icon.component';
import * as ɵngcc9 from '../../../select/select-item.component';
import * as ɵngcc10 from '../../../input-group/input-group.component';
import * as ɵngcc11 from '../../../directives/input/input.directive';
import * as ɵngcc12 from '../../../directives/button/button.directive';
import * as ɵngcc13 from '../../../buttonGroup/buttonGroup.component';
import * as ɵngcc14 from '../../common/pipes';

const _c0 = ["datePicker"];
function IgxExcelStyleDateExpressionComponent_igx_icon_3_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelement(0, "igx-icon", 12);
} if (rf & 2) {
    const ctx_r1 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("name", ctx_r1.getIconName());
} }
function IgxExcelStyleDateExpressionComponent_igx_icon_4_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "igx-icon");
    ɵngcc0.ɵɵtext(1, "filter_list");
    ɵngcc0.ɵɵelementEnd();
} }
function IgxExcelStyleDateExpressionComponent_igx_select_item_5_Template(rf, ctx) { if (rf & 1) {
    ɵngcc0.ɵɵelementStart(0, "igx-select-item", 13);
    ɵngcc0.ɵɵelement(1, "igx-icon", 12);
    ɵngcc0.ɵɵelementStart(2, "span");
    ɵngcc0.ɵɵtext(3);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const condition_r9 = ctx.$implicit;
    const ctx_r3 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("value", condition_r9)("selected", ctx_r3.isConditionSelected(condition_r9));
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("name", ctx_r3.getCondition(condition_r9).iconName);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate(ctx_r3.translateCondition(condition_r9));
} }
function IgxExcelStyleDateExpressionComponent_ng_template_8_Template(rf, ctx) { if (rf & 1) {
    const _r15 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "igx-input-group", 14, 15);
    ɵngcc0.ɵɵelementStart(2, "input", 16, 17);
    ɵngcc0.ɵɵlistener("click", function IgxExcelStyleDateExpressionComponent_ng_template_8_Template_input_click_2_listener() { ɵngcc0.ɵɵrestoreView(_r15); const openDialog_r10 = ctx.openDialog; const _r12 = ɵngcc0.ɵɵreference(1); return openDialog_r10(_r12.element.nativeElement); });
    ɵngcc0.ɵɵpipe(4, "columnFormatter");
    ɵngcc0.ɵɵpipe(5, "date");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const value_r11 = ctx.value;
    const ctx_r5 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("displayDensity", ctx_r5.displayDensity);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("placeholder", ctx_r5.inputDatePlaceholder)("value", ctx_r5.column.formatter ? ɵngcc0.ɵɵpipeBind2(4, 5, value_r11, ctx_r5.column.formatter) : ɵngcc0.ɵɵpipeBind4(5, 8, value_r11, ctx_r5.column.pipeArgs.format, undefined, ctx_r5.grid.locale))("readonly", true)("disabled", ctx_r5.expressionUI.expression.condition && ctx_r5.expressionUI.expression.condition.isUnary);
} }
function IgxExcelStyleDateExpressionComponent_button_9_Template(rf, ctx) { if (rf & 1) {
    const _r17 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "button", 18);
    ɵngcc0.ɵɵlistener("click", function IgxExcelStyleDateExpressionComponent_button_9_Template_button_click_0_listener() { ɵngcc0.ɵɵrestoreView(_r17); const ctx_r16 = ɵngcc0.ɵɵnextContext(); return ctx_r16.onRemoveButtonClick(); });
    ɵngcc0.ɵɵelementStart(1, "igx-icon");
    ɵngcc0.ɵɵtext(2, "cancel");
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r6 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("displayDensity", ctx_r6.displayDensity);
} }
function IgxExcelStyleDateExpressionComponent_igx_buttongroup_10_Template(rf, ctx) { if (rf & 1) {
    const _r22 = ɵngcc0.ɵɵgetCurrentView();
    ɵngcc0.ɵɵelementStart(0, "igx-buttongroup", 19, 20);
    ɵngcc0.ɵɵelementStart(2, "span", 21, 22);
    ɵngcc0.ɵɵlistener("keydown", function IgxExcelStyleDateExpressionComponent_igx_buttongroup_10_Template_span_keydown_2_listener($event) { ɵngcc0.ɵɵrestoreView(_r22); const ctx_r21 = ɵngcc0.ɵɵnextContext(); return ctx_r21.onLogicOperatorKeyDown($event, 0); })("click", function IgxExcelStyleDateExpressionComponent_igx_buttongroup_10_Template_span_click_2_listener($event) { ɵngcc0.ɵɵrestoreView(_r22); const ctx_r23 = ɵngcc0.ɵɵnextContext(); return ctx_r23.onLogicOperatorButtonClicked($event, 0); });
    ɵngcc0.ɵɵtext(4);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementStart(5, "span", 21, 23);
    ɵngcc0.ɵɵlistener("keydown", function IgxExcelStyleDateExpressionComponent_igx_buttongroup_10_Template_span_keydown_5_listener($event) { ɵngcc0.ɵɵrestoreView(_r22); const ctx_r24 = ɵngcc0.ɵɵnextContext(); return ctx_r24.onLogicOperatorKeyDown($event, 1); })("click", function IgxExcelStyleDateExpressionComponent_igx_buttongroup_10_Template_span_click_5_listener($event) { ɵngcc0.ɵɵrestoreView(_r22); const ctx_r25 = ɵngcc0.ɵɵnextContext(); return ctx_r25.onLogicOperatorButtonClicked($event, 1); });
    ɵngcc0.ɵɵtext(7);
    ɵngcc0.ɵɵelementEnd();
    ɵngcc0.ɵɵelementEnd();
} if (rf & 2) {
    const ctx_r7 = ɵngcc0.ɵɵnextContext();
    ɵngcc0.ɵɵproperty("multiSelection", false);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵproperty("displayDensity", ctx_r7.displayDensity)("selected", ctx_r7.expressionUI.afterOperator === 0);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate1(" ", ctx_r7.grid.resourceStrings.igx_grid_filter_operator_and, " ");
    ɵngcc0.ɵɵadvance(1);
    ɵngcc0.ɵɵproperty("displayDensity", ctx_r7.displayDensity)("selected", ctx_r7.expressionUI.afterOperator === 1);
    ɵngcc0.ɵɵadvance(2);
    ɵngcc0.ɵɵtextInterpolate1(" ", ctx_r7.grid.resourceStrings.igx_grid_filter_operator_or, " ");
} }
export class IgxExcelStyleDateExpressionComponent extends IgxExcelStyleDefaultExpressionComponent {
    get inputValuesElement() {
        return this.datePicker.getEditElement();
    }
    get inputDatePlaceholder() {
        return this.grid.resourceStrings['igx_grid_filter_row_date_placeholder'];
    }
}
IgxExcelStyleDateExpressionComponent.ɵfac = function IgxExcelStyleDateExpressionComponent_Factory(t) { return ɵIgxExcelStyleDateExpressionComponent_BaseFactory(t || IgxExcelStyleDateExpressionComponent); };
IgxExcelStyleDateExpressionComponent.ɵcmp = ɵngcc0.ɵɵdefineComponent({ type: IgxExcelStyleDateExpressionComponent, selectors: [["igx-excel-style-date-expression"]], viewQuery: function IgxExcelStyleDateExpressionComponent_Query(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵstaticViewQuery(_c0, true, IgxDatePickerComponent);
    } if (rf & 2) {
        var _t;
        ɵngcc0.ɵɵqueryRefresh(_t = ɵngcc0.ɵɵloadQuery()) && (ctx.datePicker = _t.first);
    } }, inputs: { displayDensity: "displayDensity" }, features: [ɵngcc0.ɵɵInheritDefinitionFeature], decls: 13, vars: 11, consts: [["type", "box", 3, "displayDensity", "placeholder", "overlaySettings", "onSelection"], ["dropdownConditions", ""], ["fontSet", "imx-icons", 3, "name", 4, "ngIf"], [4, "ngIf"], [3, "value", "selected", 4, "ngFor", "ngForOf"], ["mode", "dropdown", 3, "ngModel", "locale", "outlet", "ngModelChange"], ["datePicker", ""], ["igxDatePickerTemplate", ""], ["igxButton", "icon", 3, "displayDensity", "click", 4, "ngIf"], [3, "multiSelection", 4, "ngIf"], ["igxOverlayOutlet", "", 3, "pointerdown"], ["overlayOutlet", ""], ["fontSet", "imx-icons", 3, "name"], [3, "value", "selected"], ["type", "box", 3, "displayDensity"], ["dropDownTarget", ""], ["igxInput", "", "tabindex", "0", "autocomplete", "off", 3, "placeholder", "value", "readonly", "disabled", "click"], ["input", ""], ["igxButton", "icon", 3, "displayDensity", "click"], [3, "multiSelection"], ["logicOperatorButtonGroup", ""], ["igxButton", "", "tabindex", "0", "type", "button", 3, "displayDensity", "selected", "keydown", "click"], ["andButton", ""], ["orButton", ""]], template: function IgxExcelStyleDateExpressionComponent_Template(rf, ctx) { if (rf & 1) {
        ɵngcc0.ɵɵelementStart(0, "igx-select", 0, 1);
        ɵngcc0.ɵɵlistener("onSelection", function IgxExcelStyleDateExpressionComponent_Template_igx_select_onSelection_0_listener($event) { return ctx.onConditionsChanged($event); });
        ɵngcc0.ɵɵelementStart(2, "igx-prefix");
        ɵngcc0.ɵɵtemplate(3, IgxExcelStyleDateExpressionComponent_igx_icon_3_Template, 1, 1, "igx-icon", 2);
        ɵngcc0.ɵɵtemplate(4, IgxExcelStyleDateExpressionComponent_igx_icon_4_Template, 2, 0, "igx-icon", 3);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(5, IgxExcelStyleDateExpressionComponent_igx_select_item_5_Template, 4, 4, "igx-select-item", 4);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵelementStart(6, "igx-date-picker", 5, 6);
        ɵngcc0.ɵɵlistener("ngModelChange", function IgxExcelStyleDateExpressionComponent_Template_igx_date_picker_ngModelChange_6_listener($event) { return ctx.expressionUI.expression.searchVal = $event; });
        ɵngcc0.ɵɵtemplate(8, IgxExcelStyleDateExpressionComponent_ng_template_8_Template, 6, 13, "ng-template", 7);
        ɵngcc0.ɵɵelementEnd();
        ɵngcc0.ɵɵtemplate(9, IgxExcelStyleDateExpressionComponent_button_9_Template, 3, 1, "button", 8);
        ɵngcc0.ɵɵtemplate(10, IgxExcelStyleDateExpressionComponent_igx_buttongroup_10_Template, 8, 7, "igx-buttongroup", 9);
        ɵngcc0.ɵɵelementStart(11, "div", 10, 11);
        ɵngcc0.ɵɵlistener("pointerdown", function IgxExcelStyleDateExpressionComponent_Template_div_pointerdown_11_listener($event) { return ctx.onOutletPointerDown($event); });
        ɵngcc0.ɵɵelementEnd();
    } if (rf & 2) {
        ɵngcc0.ɵɵproperty("displayDensity", ctx.displayDensity)("placeholder", ctx.conditionsPlaceholder)("overlaySettings", ctx.dropDownOverlaySettings);
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵproperty("ngIf", ctx.expressionUI.expression.condition);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", !ctx.expressionUI.expression.condition);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngForOf", ctx.conditions);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngModel", ctx.expressionUI.expression.searchVal)("locale", ctx.grid.locale)("outlet", ctx.grid.outlet);
        ɵngcc0.ɵɵadvance(3);
        ɵngcc0.ɵɵproperty("ngIf", !ctx.isSingle);
        ɵngcc0.ɵɵadvance(1);
        ɵngcc0.ɵɵproperty("ngIf", !ctx.isLast);
    } }, directives: [ɵngcc1.IgxSelectComponent, ɵngcc2.IgxPrefixDirective, ɵngcc3.NgIf, ɵngcc3.NgForOf, ɵngcc4.IgxDatePickerComponent, ɵngcc5.NgControlStatus, ɵngcc5.NgModel, ɵngcc6.IgxDatePickerTemplateDirective, ɵngcc7.IgxOverlayOutletDirective, ɵngcc8.IgxIconComponent, ɵngcc9.IgxSelectItemComponent, ɵngcc10.IgxInputGroupComponent, ɵngcc11.IgxInputDirective, ɵngcc12.IgxButtonDirective, ɵngcc13.IgxButtonGroupComponent], pipes: [ɵngcc14.IgxColumnFormatterPipe, ɵngcc3.DatePipe], encapsulation: 2 });
IgxExcelStyleDateExpressionComponent.propDecorators = {
    datePicker: [{ type: ViewChild, args: ['datePicker', { read: IgxDatePickerComponent, static: true },] }],
    displayDensity: [{ type: Input }]
};
const ɵIgxExcelStyleDateExpressionComponent_BaseFactory = /*@__PURE__*/ ɵngcc0.ɵɵgetInheritedFactory(IgxExcelStyleDateExpressionComponent);
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(IgxExcelStyleDateExpressionComponent, [{
        type: Component,
        args: [{
                preserveWhitespaces: false,
                selector: 'igx-excel-style-date-expression',
                template: "<igx-select #dropdownConditions\n        type=\"box\"\n        [displayDensity]=\"displayDensity\"\n        [placeholder]=\"conditionsPlaceholder\"\n        (onSelection)=\"onConditionsChanged($event)\"\n        [overlaySettings]=\"dropDownOverlaySettings\">\n    <igx-prefix>\n        <igx-icon *ngIf=\"expressionUI.expression.condition\" fontSet=\"imx-icons\" [name]=\"getIconName()\"></igx-icon>\n        <igx-icon *ngIf=\"!expressionUI.expression.condition\">filter_list</igx-icon>\n    </igx-prefix>\n    <igx-select-item *ngFor=\"let condition of conditions\" [value]=\"condition\" [selected]=\"isConditionSelected(condition)\">\n        <igx-icon fontSet=\"imx-icons\" [name]=\"getCondition(condition).iconName\"></igx-icon>\n        <span>{{translateCondition(condition)}}</span>\n    </igx-select-item>\n</igx-select>\n\n<igx-date-picker #datePicker mode=\"dropdown\" [(ngModel)]=\"expressionUI.expression.searchVal\" [locale]=\"grid.locale\" [outlet]=\"grid.outlet\">\n    <ng-template igxDatePickerTemplate let-openDialog=\"openDialog\" let-value=\"value\">\n        <igx-input-group #dropDownTarget type=\"box\" [displayDensity]=\"displayDensity\">\n            <input #input\n                    igxInput\n                    tabindex=\"0\"\n                    (click)=\"openDialog(dropDownTarget.element.nativeElement)\"\n                    [placeholder]=\"inputDatePlaceholder\"\n                    autocomplete=\"off\"\n                    [value]=\"\n                        column.formatter\n                        ? (value | columnFormatter:column.formatter)\n                        : (value | date:column.pipeArgs.format:undefined:grid.locale)\n                    \"\n                    [readonly]=\"true\"\n                    [disabled]=\"expressionUI.expression.condition && expressionUI.expression.condition.isUnary\"/>\n        </igx-input-group>\n    </ng-template>\n</igx-date-picker>\n\n<button (click)=\"onRemoveButtonClick()\" igxButton=\"icon\" [displayDensity]=\"displayDensity\" *ngIf=\"!isSingle\" >\n    <igx-icon>cancel</igx-icon>\n</button>\n\n<igx-buttongroup #logicOperatorButtonGroup\n    *ngIf=\"!isLast\"\n    [multiSelection]=\"false\">\n    <span igxButton [displayDensity]=\"displayDensity\"\n        #andButton\n        (keydown)=\"onLogicOperatorKeyDown($event, 0)\"\n        tabindex=\"0\"\n        [selected]=\"expressionUI.afterOperator === 0\"\n        type=\"button\"\n        (click)=\"onLogicOperatorButtonClicked($event, 0)\">\n        {{ grid.resourceStrings.igx_grid_filter_operator_and }}\n    </span>\n\n    <span igxButton [displayDensity]=\"displayDensity\"\n        #orButton\n        tabindex=\"0\"\n        (keydown)=\"onLogicOperatorKeyDown($event, 1)\"\n        [selected]=\"expressionUI.afterOperator === 1\"\n        type=\"button\"\n        (click)=\"onLogicOperatorButtonClicked($event, 1)\">\n        {{ grid.resourceStrings.igx_grid_filter_operator_or }}\n    </span>\n</igx-buttongroup>\n\n<div #overlayOutlet\n     igxOverlayOutlet\n     (pointerdown)=\"onOutletPointerDown($event)\">\n</div>\n"
            }]
    }], null, { datePicker: [{
            type: ViewChild,
            args: ['datePicker', { read: IgxDatePickerComponent, static: true }]
        }], displayDensity: [{
            type: Input
        }] }); })();
